resources/assets/lib/notification-widget/item.tsx

```diff
@@ -0,0 +1,252 @@
+          <div
+            className='notification-popup-item__time'
+            dangerouslySetInnerHTML={{
+              __html: osu.timeago(item.createdAtJson),
+            }}
+          />
+        </div>
+        <div className='notification-popup-item__side-buttons'>
+          {this.renderMarkAsReadButton()}
+        </div>
+      </div>
+    </div>;
+  }
+
+  private renderCoverIcon() {
+    if (this.props.items.length === 0) {
+      return null;
+    }
+
+    const item = this.props.items[0];
+
+    if (item.name == null || item.category == null) {
+      return null;
+    }
+
+    const icons = this.props.items.length === 1
+      ? ITEM_NAME_ICONS[item.name]
+      : ITEM_CATEGORY_ICONS[item.category];
+
+    if (icons == null) {
+      return null;
+    }
+
+    return icons.map((icon) => {
+      return <div key={icon} className='notification-popup-item__cover-icon'>
+        <span className={icon} />
+      </div>;
+    });
+  }
+
+  private renderMarkAsReadButton() {
+    if (this.props.items[0].id < 0) {
+      return null;
+    }
+
+    if (this.state.markingAsRead) {
+      return <div className='notification-popup-item__read-button'>
+          <Spinner />
+        </div>;
+    } else {
+      return <button
+          type='button'
+          className='notification-popup-item__read-button'
+          onClick={this.markRead}
+        >
+          <span className='fas fa-times' />
+        </button>;
+    }
+  }
+
+  private renderMessage() {
+    if (this.props.items.length === 0) {
+      return null;
+    }
+
+    const item = this.props.items[0];
+    let message: string;
+
+    const replacements = {
+      title: item.details.title,
+      username: item.details.username,
+    };
+
+    if (this.props.items.length === 1) {
+      const key = `notifications.item.${item.objectType}.${item.category}.${item.name}`;
+
+      if (item.name === 'legacy_pm') {
+        message = osu.transChoice(key, item.details.count, replacements);
+      } else {
+        message = osu.trans(key, replacements);
+      }
+    } else {
+      message = osu.transChoice(`notifications.message_multi`, this.props.items.length, replacements);
+    }
+
+    return <a href={this.url()} className='notification-popup-item__message'>
+      {message}
+    </a>;
+  }
+
+  private markRead = () => {
+    this.setState({ markingAsRead: true });
+    const ids = this.props.items.map((i) => i.id);
+
+    this.props.worker.sendMarkRead(ids)
+    .fail(() => this.setState({ markingAsRead: false }));
+  }
+
+  private url() {
+    if (this.props.items.length === 0) {
+      return;
+    }
+
+    const item = this.props.items[0];
+
+    if (item.name === 'legacy_pm') {
+      return '/forum/ucp.php?i=pm&folder=inbox';
+    }
+
+    let route: string = '';
+    let params: any;
+
+    if (this.props.items.length === 1) {
+      switch (item.name) {
+        case 'beatmapset_discussion_post_new':
+          route = 'beatmap-discussions.show';
```

Comment:
I wonder if these URLs should be generated server-side to prevent the game-client having to generate all these too?... Having said that, how will the game client handle clicking on these? I imagine for beatmapset/etc it should probably display in-game and not open the website?